cxml/xml.go

```diff
@@ -0,0 +1,222 @@
+/*
+Copyright 2015 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package cxml
+
+import (
+	"encoding/xml"
+	"io"
+	"strings"
+
+	"github.com/gravitational/trace"
+)
+
+// TransofrmFunc is a function that can transform incoming token
+// into a series of outgoing tokens when traversing XML tree
+type TransformFunc func(parents *NodeList, in xml.Token) []xml.Token
+
+// ConditionFunc is accepted by some transform function generators
+// so they could be applied conditionaly
+type ConditionFunc func(parents *NodeList, in xml.Token) bool
+
+// TransformXML parses the XML tree, traverses it and calls TransformFunc
+// on each XML token, writing the output to the writer, resulting in a
+// transformed XML tree
+func TransformXML(r io.Reader, w io.Writer, fn TransformFunc, indent bool) error {
```

Comment:
Better to accept `xml.Decoder`/`xml.Encoder` - additional work performed in 38-42 runs counter to principle of single responsibility. This will make `TranformXML` more composable and simpler:
```go
TransformXML(r *xml.Decoder, w *xml.Encoder, fn TransformFunc) bool
```